### ISSUE TITLE
Add environment variable flag to python

### ISSUE THREAD

**Issue by AdamSimpson** [olcf/wraprun#5]
Add flag to enable per process/color environment variables to be set. List of environment variables should be passed into WRAPRUN_FILE as fourth column in the format "key1=value2;key2=value2".

**Comment by avmarchenko** [comment]
Hello, was there ever any progress on this? I glanced at the code and it seems that (if using Python) when calling and instance of the Wraprun class's launch method there is an os.environ.update call (using the class instance's env (_env) - a dict of env var names/values I presume)? If I need to set a per task env variable can I accomplish it using Python or does launch set the global environment only? Can it be done via command line (without Python)? Could a hack be made to _subprocess_args to enable env defintions prior to aprun calls? Is this not possible because split.c doesn't know about env vars?

**Comment by AdamSimpson** [comment]
The underlying C library supports per-process environment variables although there is no easy way to set them currently. Paging @mpbelhorn to see what he thinks.

**Comment by mpbelhorn** [comment]
Per-process envvars can probably be done, but I don't think setting them in the `Wraprun.launch` environment within python is useful. That's just the environment in which aprun is called on the batch node. I think we'd need to write these variables to the `WRAPRUN_FILE` and parse/set them from the C library for each process.

**Comment by AdamSimpson** [comment]
The C library parses the fourth column in the `WRAPRUN_FILE` for environment variables. The column must be formated as `key1=value2;key2=value2`

**Comment by avmarchenko** [comment]
The python Wraprun class writes the ``WRAPRUN_FILE`` correct? Does that mean modifications only need to be made to the py side of things?

**Comment by mpbelhorn** [comment]
Yes, all the additions would need to be made in the python code. @AdamSimpson where are envvars currently added to the 4th `WRAPRUN_FILE` column? This looks like it was added in a very early commit. Is this vestigial code from when we were trying to pass all the rank parameters via envvars?

If that column is not already used for anything, then at a minimum, the `wraprun.task.Rank` object needs to be updated to have it's `FILE_CONTENT` tuple have an additional item for envvars that is by default an empty string. An argument would need to be added to  `wraprun.options.GroupOptions` to parse input strings as Adam described, and [the rank instantiation](https://github.com/olcf/wraprun/blob/master/python/wraprun/task.py#L163-L165) and [ctor](https://github.com/olcf/wraprun/blob/master/python/wraprun/task.py#L29-L38) would need to add a keyword arg to set the envar string obtained from the option parser. I'm not immediately sure if any other changes would need to be made.

If the column *is* already used then we will need to make sure whatever is currently put there is preserved and merged with what [`wraprun.api._update_file`](https://github.com/olcf/wraprun/blob/master/python/wraprun/api.py#L203) will write there.

**Comment by AdamSimpson** [comment]
Ah good call, it's the forth column not the third. All of the logic appears to be present in libsplit to handle a fourth column of environment variable key-values if present.

**Comment by mpbelhorn** [comment]
Yeah, I'll start developing this feature. I don't think it'll take too long to implement.

**Comment by mpbelhorn** [comment]
Ok, I haven't tested it yet but the commit I just pushed appears to produce the correct form for the WRAPRUN_FILE lines:

```
$ export WRAPRUN_PRELOAD=none
$ wraprun --w-debug -n 3 --w-env 'key1=herp;key2=derp,key3=perp,,key4=foo' exe
BEGIN WRAPRUN DEBUGGING INFO
 Aprun call signature:
   aprun -b -n 12 exe

 Environment variables:
   LD_PRELOAD=none
   WRAPRUN_FILE=/autofs/nccs-svm1_home2/belhorn/development/wraprun/python/wraprun_H1PwrC.tmp
   W_IGNORE_SEGV=1
   W_REDIRECT_OUTERR=1
   W_UNSET_PRELOAD=1

 Internal state:
   Wraprun(ranks=12, colors=4)

 Tempfile contents:
   00|0 ./ unnamed.31601_w15.0 key1=herp;key2=derp
   01|0 ./ unnamed.31601_w15.0 key1=herp;key2=derp
   02|0 ./ unnamed.31601_w15.0 key1=herp;key2=derp
   03|1 ./ unnamed.31601_w15.1 key3=perp
   04|1 ./ unnamed.31601_w15.1 key3=perp
   05|1 ./ unnamed.31601_w15.1 key3=perp
   06|2 ./ unnamed.31601_w15.2 
   07|2 ./ unnamed.31601_w15.2 
   08|2 ./ unnamed.31601_w15.2 
   09|3 ./ unnamed.31601_w15.3 key4=foo
   10|3 ./ unnamed.31601_w15.3 key4=foo
   11|3 ./ unnamed.31601_w15.3 key4=foo
END WRAPRUN DEBUGGING INFO
```

Just need to deploy this on Titan and run the tests. I pushed it to the development branch so I can write a spack package to install it for testing.

**Comment by AdamSimpson** [comment]
nice work!

**Comment by avmarchenko** [comment]
Thanks for doing all of this so quickly. Once pushed to Titan, let me know. I have a small real life test that I can try this out on. Thanks again.

**Comment by mpbelhorn** [comment]
I'm actually going to be out of the office for the rest of the day and won't have time to deploy this until tomorrow morning.

**Comment by mpbelhorn** [comment]
@avmarchenko Ok, it was failing some test earlier but I think I've worked out the bug. What is your preferred programming environment PrgEnv-* and compiler version?  I'll deploy a development build of it that you can use for that compiler and you can try it out before we cut a new release.

---

EDIT: There is already one dev build installed in my personal testing area that should work with PrgEnv-gnu and gcc/4.9.3  through gcc/5.3.0. To use it, you can call:

```bash
# from a clean login environment:
module swap PrgEnv-pgi PrgEnv-gnu
module use /sw/.testing/belhorn/titan/.swci/1-compute/share/spack/modules/20170104/cray-CNL-interlagos
module load wraprun-dev-gcc-5.3.0-5kjpvpa py-pyyaml-3.11-gcc-5.3.0-sbwdokw

# test it runs with
wraprun -h
```

That build does not require you to load a specific python and is hardcoded to use a personal build of python2.7.12

**Comment by avmarchenko** [comment]
I use pgi and intel primarily. I can test on pgi since that is also the default. Thanks so much for such a rapid implementation.

**Comment by mpbelhorn** [comment]
Ok, I've installed a default pgi/16.10.0 development build in my testing area. To use it:

```
# from a clean login environment:
module use /sw/.testing/belhorn/titan/.swci/1-compute/share/spack/modules/20170104/cray-CNL-interlagos
module load py-pyyaml-3.11-pgi-16.10.0-ily5c7s wraprun-dev-pgi-16.10.0-byhkyon  

# test it runs with
wraprun -h
```

It passes a number of simple tests and appears to be setting envvars correctly in each MPI rank's runtime environment. Updated documentation for it is under "Per-task environment variables" in the development branch [README.md](https://github.com/olcf/wraprun/tree/development). Let me know if something goes wrong. If it all works we'll cut a v0.2.6 release for all the default compilers.

**Comment by avmarchenko** [comment]
Sorry for the delay. I think my testing scheme must be incorrect because things are not working as expected. I have a directory ``test`` that looks like this:
```
1/
2/
1.inp
2.inp
test.sub
```

The ``inp`` files contain parameters for the executable and the corresponding directories are where scratch data for each calculation is written. The batch script, ``test.sub``, contains:

```
...
module use /sw/.testing/belhorn/titan/.swci/1-compute/share/spack/modules/20170104/cray-CNL-interlagos
module load py-pyyaml-3.11-pgi-16.10.0-ily5c7s wraprun-dev-pgi-16.10.0-byhkyon
module load python

wraprun -n n --w-env ENVVAR=value1 executable 1.inp > 1.out : \
              -n n --w-env ENVVAR=value2 executable 2.inp > 2.out
```

Only ``2.out`` contains any output from the calculation and it shows that ``ENVVAR=value2`` was not respected. Any pointers would be appreciated.

**Comment by mpbelhorn** [comment]
@avmarchenko You will want to put the envar settings in quotes and it is not possible to have shell redirection in between each MPMD groups separated by ":". The redirection should not be needed anyway because wraprun outputs stderr/stdout for each group to files named:

```
${JOBNAME}.${JOBID}_w${INSTANCE}.${TASKID}.out
${JOBNAME}.${JOBID}_w${INSTANCE}.${TASKID}.err
```

The name of which can be overridden per group by the `--w-oe BASENAME` flag.

Try the following format:

```
...
module use /sw/.testing/belhorn/titan/.swci/1-compute/share/spack/modules/20170104/cray-CNL-interlagos
module load py-pyyaml-3.11-pgi-16.10.0-ily5c7s wraprun-dev-pgi-16.10.0-byhkyon
module load python

wraprun -n n --w-env "ENVVAR=value1" --w-oe 1 executable 1.inp : \
              -n n --w-env "ENVVAR=value2" --w-oe 2 executable 2.inp
```

which should write output files 1.out, 1.err, 2.out, 2.err in the current working directory.

**Comment by avmarchenko** [comment]
If the executable is added by sourcing a ``.sh`` file in an earlier part of the script, should it be available to wraprun?

**Comment by avmarchenko** [comment]
I tried the following but without success; the executable doesn't respect the environment variable.

```
...
wraprun -n 16 --w-env "VARNAME=/explicitly/defined/path1" --w-oe 1 executable 1.inp : \
              -n 16 --w-env "VARNAME=/explicitly/defined/path2" --w-oe 2 executable 2.inp
```

The executable is an MPI application that I would normally run like ``aprun exec... input``. The variable value in the output is being read from the compiled defaults of the executable rather than the environment. If I do a test where I set the environment variable and make an aprun call as above the environment variable is respected.

**Comment by robertdfrench** [comment]
I'm having the same issue on Tiny Titan, will this be resolved soon?
